backend/src/composite/mod.rs

```diff
@@ -247,52 +374,96 @@ impl<'a, F: FieldElement> Backend<'a, F> for CompositeBackend<'a, F> {
-                    .iter()
-                    .map(|(_, witness)| witness.len())
-                    .unique()
-                    .exactly_one()
-                    .expect("All witness columns of a machine must have the same size");
-                let machine_data = machine_data
-                    .get(&size)
-                    .expect("Machine does not support the given size");
-                let witgen_callback = witgen_callback.clone().with_pil(machine_data.pil.clone());
-
-                log::info!("== Proving machine: {} (size {})", machine, size);
-                log::debug!("PIL:\n{}", machine_data.pil);
-
-                let proof = machine_data.backend.prove(&witness, None, witgen_callback);
-
-                match proof {
-                    Ok(inner_proof) => {
-                        log::info!(
+        // We use scoped threads to be able to share non 'static references.
+        thread::scope(|scope| {
+            let mut proofs_status = self
+                .machine_data
+                .iter()
+                .map(|(machine, machine_data)| {
+                    let (witness, size) =
+                        process_witness_for_machine(machine, machine_data, witness);
+
+                    let inner_machine_data = machine_data
+                        .get(&size)
+                        .expect("Machine does not support the given size");
+
+                    log::info!("== Proving machine: {} (size {})", machine, size);
+                    log::debug!("PIL:\n{}", inner_machine_data.pil);
+
+                    (
+                        sub_prover::run(scope, &inner_machine_data.backend, witness),
+                        machine_data,
+                        size,
+                    )
+                })
+                .collect::<Vec<_>>();
+
+            let mut proof_results = Vec::new();
+            let mut witness = Cow::Borrowed(witness);
+            for stage in 0.. {
+                if proofs_status.is_empty() {
+                    break;
+                }
+
+                // Filter out proofs that have completed and accumulate the
+                // challenges.
+                let mut challenges = BTreeMap::new();
+                let waiting_provers = std::mem::take(&mut proofs_status)
+                    .into_iter()
+                    .filter_map(|(status, machine_data, size)| match status {
+                        sub_prover::RunStatus::Completed(result) => {
+                            proof_results.push((result, size));
+                            None
+                        }
+                        sub_prover::RunStatus::Challenged(sub_prover, c) => {
+                            // Accumulate the challenges
+                            accumulate_challenges(&mut challenges, c);
+                            Some((sub_prover, machine_data))
+                        }
+                    })
+                    .collect::<Vec<_>>();
```

Comment:
```suggestion
                    .collect::<Vec<_>>();
                    
                if proofs_status.is_empty() {
                    break;
                }
```
Should be moved from above, so that we don't run the second-phase witness generation when all the proofs are done. This is also causing some tests to fail.

Reply:
There is something else wrong. `proof_status` is always empty after `std::mem::take(&mut proof_status)`.